packages/xo-web/src/xo-app/settings/audit/index.js

```diff
@@ -1,17 +1,165 @@
+          {getIntegrityErrorRender({ nValid, errorData })}
+          <p>{_('auditSaveFingerprintInErrorInfo')}</p>
+        </div>
+      ) : (
+        <p>{_('auditSaveFingerprintInfo')}</p>
+      )}
+      <p className='input-group mt-1'>
+        <input className='form-control' value={fingerprint} disabled />
+        <span className='input-group-btn'>
+          <Tooltip content={_('auditCopyFingerprintToClipboard')}>
+            <CopyToClipboard text={fingerprint}>
+              <Button>
+                <Icon icon='clipboard' />
+              </Button>
+            </CopyToClipboard>
+          </Tooltip>
+        </span>
+      </p>
+    </div>
+  ).catch(noop)
+
+const openIntegrityFeedbackModal = error =>
+  chooseAction({
+    icon: 'diagnosis',
+    title: _('auditCheckIntegrity'),
+    body:
+      error !== undefined ? (
+        getIntegrityErrorRender(error)
+      ) : (
+        <p className='text-success'>
+          {_('auditIntegrityVerified')} <Icon icon='success' />
+        </p>
+      ),
+    buttons: [
+      {
+        btnStyle: 'success',
+        label: _('auditGenerateNewFingerprint'),
+      },
+    ],
+  }).then(
+    () => true,
+    () => false
+  )
+
+const FingerPrintModalBody = injectIntl(
+  ({ intl: { formatMessage }, onChange, value }) => (
+    <div>
+      <p>{_('auditEnterFingerprintInfo')}</p>
+      <div className='form-group'>
+        <input
+          className='form-control'
+          onChange={onChange}
+          pattern='[^|]+\|[^|]+'
+          placeholder={formatMessage(messages.auditEnterFingerprint)}
+          value={value}
+        />
+      </div>
+    </div>
+  )
+)
+
+const DEFAULT_HASH = 'nullId|nullId'
+const openFingerprintPromptModal = () =>
+  form({
+    render: ({ onChange, value }) => (
+      <FingerPrintModalBody onChange={onChange} value={value} />
+    ),
+    header: (
+      <span>
+        <Icon icon='diagnosis' /> {_('auditCheckIntegrity')}
+      </span>
+    ),
+  }).then((value = '') => {
+    value = value.trim()
+    return value !== '' ? value : DEFAULT_HASH
+  }, noop)
+
+const checkIntegrity = async () => {
+  const fingerprint = await openFingerprintPromptModal()
+  if (fingerprint === undefined) {
+    return
+  }
+
+  const [oldest, newest] = fingerprint.split('|')
+  const error = await checkAuditRecordsIntegrity(oldest, newest).then(
+    noop,
+    error => {
+      const isRecordMissing = missingAuditRecord.is(error)
+      if (isRecordMissing || alteredAuditRecord.is(error)) {
+        return {
+          nValid: error.nValid,
+          errorData: {
+            id: error.id,
+            reason: isRecordMissing
+              ? missingAuditRecord.message
+              : alteredAuditRecord.message,
+          },
+        }
+      }
+      throw error
+    }
+  )
+
+  const shouldGenerateFingerprint = await openIntegrityFeedbackModal(error)
+  if (shouldGenerateFingerprint) {
+    openGeneratedFingerprintModal(await generateAuditFingerprint(newest))
```

Comment:
Don't break the promise chain.

Reply:
It will not break, the error is properly handled.

Reply:
If you don't `return` or `await` the promise, the chain is broken.